File: src/engine/ecs/EntityManager.cpp
#include "EntityManager.hpp"

namespace voxel {

EntityManager::EntityManager(int32_t capacity)
  : m_capacity(capacity), m_generation(capacity, 0), m_freeIndices(capacity) {
  for (int32_t i = capacity - 1; i >= 0; --i) {
    m_freeIndices[m_freeHead++] = i;
  }
}

auto EntityManager::allocate() -> int32_t {
  if (m_freeHead == 0) {
    throw std::runtime_error("EntityManager capacity exhausted");
  }
  int32_t idx = m_freeIndices[--m_freeHead];
  uint8_t gen = m_generation[idx];
  ++m_liveCount;
  return (static_cast<int32_t>(gen) << ENTITY_GEN_SHIFT) | idx;
}

void EntityManager::destroy(int32_t id) {
  int32_t idx = id & ENTITY_INDEX_MASK;
  uint8_t gen = static_cast<uint8_t>(id >> ENTITY_GEN_SHIFT);
  if (gen != m_generation[idx]) return; // already destroyed
  m_generation[idx] = (gen + 1) & 0xFF;
  m_freeIndices[m_freeHead++] = idx;
  --m_liveCount;
}

auto EntityManager::isAlive(int32_t id) const -> bool {
  int32_t idx = id & ENTITY_INDEX_MASK;
  return static_cast<uint8_t>(id >> ENTITY_GEN_SHIFT) == m_generation[idx];
}

} // namespace voxel

File: src/engine/ecs/EntityManager.ts
const INDEX_MASK = 0x00ff_ffff;
const GEN_SHIFT = 24;

export class EntityManager {
  private readonly generation: Uint8Array;
  private readonly freeIndices: Int32Array;
  private freeHead = 0;
  private liveCount = 0;

  constructor(readonly capacity: number = 1 << 18) {
    this.generation = new Uint8Array(capacity);
    this.freeIndices = new Int32Array(capacity);
    for (let i = capacity - 1; i >= 0; i--) this.freeIndices[this.freeHead++] = i;
  }

  allocate(): number {
    if (this.freeHead === 0) throw new Error("EntityManager capacity exhausted");
    const idx = this.freeIndices[--this.freeHead];
    const gen = this.generation[idx];
    this.liveCount++;
    return (gen << GEN_SHIFT) | idx;
  }

  destroy(id: number): void {
    const idx = id & INDEX_MASK;
    const gen = id >>> GEN_SHIFT;
    if (gen !== this.generation[idx]) return;
    this.generation[idx] = (gen + 1) & 0xff;
    this.freeIndices[this.freeHead++] = idx;
    this.liveCount--;
  }

  isAlive(id: number): boolean {
    const idx = id & INDEX_MASK;
    return (id >>> GEN_SHIFT) === this.generation[idx];
  }

  static indexOf(id: number): number {
    return id & INDEX_MASK;
  }

  get count(): number {
    return this.liveCount;
  }
}
